Use `rmake` for `windows-` run-make tests

ChrisDenton opened the pull request:
Convert some Makefile tests to recipes.

I renamed "issue-85441" to "windows-ws2_32" as I think it's slightly more descriptive. EDIT: `llvm-readobj` seems to work for reading DLL imports so I've used that instead of `objdump`.

cc #121876

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9254318453/job/25455860158) [(plain)](https://github.com/rust-lang/rust/commit/dce71a7f58c6236bd232a21a149145d0496d2a90/checks/25455860158/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:a5ac7e51b41094c92402da3b24376905380afc29)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.419   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.431 Collecting boolean-py==4.0
#13 0.434   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.445 Collecting chardet==5.1.0
---
#13 3.431 Building wheels for collected packages: reuse
#13 3.432   Building wheel for reuse (pyproject.toml): started
#13 3.776   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.777   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.777   Stored in directory: /tmp/pip-ephem-wheel-cache-j8j32qll/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.780 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.802   Attempting uninstall: setuptools
#13 3.803     Found existing installation: setuptools 59.6.0
#13 3.804     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.804     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.804     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.476 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.476 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.978 Collecting virtualenv
#13 5.015   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.080      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 62.8 MB/s eta 0:00:00
#13 5.127 Collecting platformdirs<5,>=3.9.1
#13 5.130   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.147 Collecting distlib<1,>=0.3.7
#13 5.150   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.187 Collecting filelock<4,>=3.12.2
#13 5.190   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.190   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.278 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.437 Successfully installed distlib-0.3.8 filelock-3.14.0 platformdirs-4.2.2 virtualenv-20.26.2
#13 DONE 5.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      184256 kB
DirectMap2M:     8204288 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/b582f807fae230b22ac126ff1d8a13262bb099ba/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-b582f807fae230b22ac126ff1d8a13262bb099ba-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished `release` profile [optimized] target(s) in 24.60s
##[endgroup]
fmt check
tidy check
tidy error: Makefile `/checkout/tests/run-make/issue-85441/Makefile` no longer exists and should be removed from the exclusions in `src/tools/tidy/src/allowed_run_make_makefiles.txt`, you can run --bless to update the allow list
tidy error: Makefile `/checkout/tests/run-make/windows-binary-no-external-deps/Makefile` no longer exists and should be removed from the exclusions in `src/tools/tidy/src/allowed_run_make_makefiles.txt`, you can run --bless to update the allow list
tidy error: Makefile `/checkout/tests/run-make/windows-safeseh/Makefile` no longer exists and should be removed from the exclusions in `src/tools/tidy/src/allowed_run_make_makefiles.txt`, you can run --bless to update the allow list
tidy error: Makefile `/checkout/tests/run-make/windows-spawn/Makefile` no longer exists and should be removed from the exclusions in `src/tools/tidy/src/allowed_run_make_makefiles.txt`, you can run --bless to update the allow list
tidy error: Makefile `/checkout/tests/run-make/windows-subsystem/Makefile` no longer exists and should be removed from the exclusions in `src/tools/tidy/src/allowed_run_make_makefiles.txt`, you can run --bless to update the allow list
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.0)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 40.4 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 34.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 34.5 MB/s eta 0:00:00
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 133.6 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:00:57
  local time: Mon May 27 11:57:06 UTC 2024
  network time: Mon, 27 May 2024 11:57:06 GMT
```

</details>

jieyouxu commented on tests/run-make/windows-safeseh/rmake.rs:
```diff
@@ -0,0 +1,15 @@
+fn main() {
+    // Ensure that LLD can link when an .rlib contains a synthetic object
+    // file referencing exported or used symbols.
+    rustc().input("foo.rs").arg("-Clinker=rust-lld").run();
```
Question: should we add a `linker()` helper method to `Rustc`? Something simple like

```rs
impl Rustc {
    fn linker(&mut self, linker: &str) -> &mut Self {
        self.cmd.arg(format!("-Clinker={linker}"));
        self
    }
}
```

ChrisDenton replied:
Makes sense to me!

jieyouxu commented on tests/run-make/windows-subsystem/rmake.rs:
Suggestion: maybe add a comment explain what this is trying to test? Is this checking that `#![windows_subsystem = "..."]` can successfully compile and run?

ChrisDenton replied:
Yeah, I *think* it's just testing it compiles on all platforms. Tbh I'm not entirely sure why this is a run-make test.

jieyouxu replied:
I think this can be two separate UI tests (basically just `windows.rs` and `console.rs` each being their own UI tests but `//@ run-pass`?)

```rs
// windows.rs
//@ run-pass
#![windows_subsystem = "windows"]
fn main() {}
```

ChrisDenton replied:
Yep. I'll try moving them.

jieyouxu reviewed (approved):
Thanks for the test comments, they are very informative :3
I left two suggestions, r=me with or without addressing them.

jieyouxu commented on tests/run-make/windows-ws2_32/rmake.rs:
```diff
@@ -0,0 +1,14 @@
+    rustc().input("empty.rs").run();
+    let empty = tmp_dir().join("empty.exe");
+    let output = llvm_readobj().input(empty).coff_imports().command_output();
+    assert!(output.status.success());
```
Reminder for myself: cc https://github.com/rust-lang/rust/issues/125617 we should add run() that asserts success for llvm-readobj as well, so it's less of a footgun to forgor checking cmd status

ChrisDenton replied:
Oh, wait there is a run I think. I missed that.

jieyouxu replied:
Yeah, I think we probably need to un-provide the `command_output` helper because it's probably almost never the right helper to use

ChrisDenton replied:
Sounds helpful. I should probably also read the docs for the support lib.

ChrisDenton replied:
Oh, that gives me a thought: it might be nice to make `./x doc run-make-support --open` work.

jieyouxu replied:
Yes, that would indeed be very useful

ChrisDenton commented:
Ok, I'll stop fiddling now 😛There are maybe some more improvements to be made but I'll leave that to follow-ups. Like, maybe it would make sense to use the `object` crate for enumerating imports rather than `llvm_readobj`.

@bors r=jieyouxu

bors commented:
:pushpin: Commit ad5dce55e620db3f992421debafe66194e5210c9 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu ad5dce55e620db3f992421debafe66194e5210c9 -->
<!-- homu: {"type":"Approved","sha":"ad5dce55e620db3f992421debafe66194e5210c9","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
i686-windows-gnu funtimes ahead?

@bors rollup=iffy

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9277541451/job/25526928982) [(plain)](https://github.com/rust-lang-ci/rust/commit/4b429647ed06a4436bea7faab515d4d5091b0f60/checks/25526928982/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling ammonia v4.0.0
[RUSTC-TIMING] html5ever test:false 2.302
[RUSTC-TIMING] toml test:false 5.259
   Compiling mdbook v0.4.40
##[error]The runner has received a shutdown signal. This can happen when the runner service is stopped, or a manually started runner is canceled.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9277541451/job/25528118184)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9277541451/job/25528118184","builder_name":"checks-actions"} -->

ChrisDenton commented:
@bors retry The runner has received a shutdown signal

bors commented:
:hourglass: Testing commit ad5dce55e620db3f992421debafe66194e5210c9 with merge fb51598d9a6320aa9bee7b5319eea3c9560ed18f...
<!-- homu: {"type":"BuildStarted","head_sha":"ad5dce55e620db3f992421debafe66194e5210c9","merge_sha":"fb51598d9a6320aa9bee7b5319eea3c9560ed18f"} -->

rust-log-analyzer commented:
The job **`x86_64-mingw`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9279132958/job/25531221728) [(plain)](https://github.com/rust-lang-ci/rust/commit/fb51598d9a6320aa9bee7b5319eea3c9560ed18f/checks/25531221728/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [run-make] tests\run-make\windows-binary-no-external-deps stdout ----

error: rmake recipe failed to complete
status: exit code: 1
command: "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\run-make\\windows-binary-no-external-deps\\windows-binary-no-external-deps\\rmake.exe"
--- stderr -------------------------------
command failed at line 13
command failed at line 13
"C:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\bin\\rustc.exe" "--out-dir" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\run-make\\windows-binary-no-external-deps\\windows-binary-no-external-deps" "-L" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\run-make\\windows-binary-no-external-deps\\windows-binary-no-external-deps" "hello.rs"
output status: `exit code: 0xc0000135`
=== STDOUT ===


=== STDERR ===
------------------------------------------
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9279132958/job/25533512360)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9279132958/job/25533512360","builder_name":"checks-actions"} -->

ChrisDenton commented:
Ok so the no external deps test is failing on mingw due to a missing dll (but it succeeds for msvc). The weird thing is it was obviously succeeding in the Makefile version. Which perhaps suggest make or mingw bash is doing something behind our backs so we weren't testing what we thought we were testing. Either that or I missed something else.

I'll investigate tomorrow.

jieyouxu commented:
> Ok so the no external deps test is failing on mingw due to a missing dll (but it succeeds for msvc). The weird thing is it was obviously succeeding in the Makefile version. Which perhaps suggest make or mingw bash is doing something behind our backs so we weren't testing what we thought we were testing. Either that or I missed something else.
> 
> I'll investigate tomorrow.

`run()` sets extra env vars for the executable, e.g.

https://github.com/rust-lang/rust/blob/751691271d76b8435559200b84d1947c2bd735bd/src/tools/run-make-support/src/run.rs#L27-L34

maybe that is why in the original Makefile they used

```make
$(TMPDIR)/hello.exe
```

instead of

```make
$(call RUN,hello)
```

Not entirely sure.

bors commented:
:umbrella: The latest upstream changes (presumably #125691) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ChrisDenton commented:
I had a look at [the original issue](https://github.com/rust-lang/rust/issues/71061) that provoked adding this test. It seems we only need to be testing the final binary, not rustc itself. That does make things easier.

I still need to rebase and squish. And while I'm at it, I might rewrite the `windows-ws2_32` test to use `object`. I think it'll be easy enough.

ChrisDenton committed:
Move run-make windows_subsystem tests to ui tests

ChrisDenton committed:
Rename run-make/issue-85441 and convert to rmake

ChrisDenton committed:
Convert run-make/windows-spawn to rmake

ChrisDenton committed:
Convert run-make/windows-safeseh to rmake

ChrisDenton committed:
Convert windows-binary-no-external-deps to rmake

ChrisDenton committed:
Remove Makefiles from allowed_run_make_makefiles

ChrisDenton commented:
I've rebased and made the commits easier to navigate. There are two changed tests.

https://github.com/rust-lang/rust/pull/125613/commits/5ec0c002edf05d7a5a3596c26f30ffa258b580bb fixes the failure as discussed above by only setting `PATH` when running the final binary

For https://github.com/rust-lang/rust/pull/125613/commits/f08e00f3d56645c9bbd4d2ebfbabe257b56b837b I switched to using `object` rather than depending on llvm-readobj and searching the command output. I also added an extra test case to make sure the actual test works. The problem with depending on the absence of something is that it could be absent for a number of reasons (e.g. the test is buggy or the standard library switches to another DLL) so testing the test still works is useful, I think.

jieyouxu commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit f34bbdecc4cd8e39470cbe2cefe96d4aed351da9 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu f34bbdecc4cd8e39470cbe2cefe96d4aed351da9 -->
<!-- homu: {"type":"Approved","sha":"f34bbdecc4cd8e39470cbe2cefe96d4aed351da9","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit f34bbdecc4cd8e39470cbe2cefe96d4aed351da9 with merge e9b7aa08f7cd890dcebd36c198193707d5cd3770...
<!-- homu: {"type":"BuildStarted","head_sha":"f34bbdecc4cd8e39470cbe2cefe96d4aed351da9","merge_sha":"e9b7aa08f7cd890dcebd36c198193707d5cd3770"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9290942114/job/25573388625)
Approved by: jieyouxu
Pushing e9b7aa08f7cd890dcebd36c198193707d5cd3770 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jieyouxu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9290942114/job/25573388625"},"merge_sha":"e9b7aa08f7cd890dcebd36c198193707d5cd3770"} -->

rust-timer commented:
Finished benchmarking commit (e9b7aa08f7cd890dcebd36c198193707d5cd3770): [comparison URL](https://perf.rust-lang.org/compare.html?start=a83f933a9da258cf037e3cab37cd486bfd861a7d&end=e9b7aa08f7cd890dcebd36c198193707d5cd3770&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=a83f933a9da258cf037e3cab37cd486bfd861a7d&end=e9b7aa08f7cd890dcebd36c198193707d5cd3770&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a83f933a9da258cf037e3cab37cd486bfd861a7d&end=e9b7aa08f7cd890dcebd36c198193707d5cd3770&stat=max-rss)
<details>
<summary>Results (primary -0.8%, secondary -3.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.2%  | [3.2%, 3.2%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-2.6%, -1.5%] | 3     |
| Improvements ✅ <br /> (secondary) | -3.6% | [-4.3%, -3.1%] | 3     |
| All ❌✅ (primary)                 | -0.8% | [-2.6%, 3.2%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a83f933a9da258cf037e3cab37cd486bfd861a7d&end=e9b7aa08f7cd890dcebd36c198193707d5cd3770&stat=cycles:u)
<details>
<summary>Results (primary -2.5%, secondary -3.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.5% | [-3.1%, -1.9%] | 6     |
| Improvements ✅ <br /> (secondary) | -3.0% | [-3.9%, -2.4%] | 14    |
| All ❌✅ (primary)                 | -2.5% | [-3.1%, -1.9%] | 6     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=a83f933a9da258cf037e3cab37cd486bfd861a7d&end=e9b7aa08f7cd890dcebd36c198193707d5cd3770&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 667.715s -> 667.139s (-0.09%)
**Artifact size**: 318.83 MiB -> 318.84 MiB (0.00%)
